coredns/endpointslice/map.go

```diff
@@ -111,6 +122,41 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
 		return
 	}
 
+	// TODO: Only do this look up if globalnet enabled to reduce unnessary call to Kube API
+	var localSlice *discovery.EndpointSlice
+	if cluster == m.localClusterID {
+		clientSet, err := kubernetes.NewForConfig(m.kubeConfig)
+		if err != nil {
+			klog.Warningf("Error creating client set %#v", err)
+			return
+		}
+
+		retryErr := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+			return !apierrors.IsNotFound(err)
+		}, func() error {
+			localSlices, err := clientSet.DiscoveryV1().EndpointSlices(es.Labels[constants.LabelSourceNamespace]).List(context.TODO(), metav1.ListOptions{
+				LabelSelector: labels.Set(map[string]string{
+					constants.KubernetesServiceName: es.Labels[constants.MCSLabelServiceName],
+				}).String(),
+			})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			} else if err != nil || len(localSlices.Items) != 1 {
+				klog.V(log.DEBUG).Infof("Unable to find local endpoint slice for service %s/%s", es.Labels[constants.LabelSourceNamespace], es.Labels[constants.MCSLabelServiceName])
+				return nil
```

Comment:
For the case where there's more than one `EndpointSlice` found, this log message isn't actually  accurate. Also we log to debug here as opposed to error on L152. However, do we need to fail in this case, ie can't we just take the first one? What would be the case where there might be more than one?

For the case where the returned list is empty, we could return a `NotFound` error and let if fall through to L152, ie 

```Go
  apierrors.NewNotFound(schema.GroupResource{
		Group:    discovery.SchemeGroupVersion.Group,
		Resource: "endpointslices",
	}, "")
```

Reply:
I think taking the first one should be fine, if there ever is more than one they should all point to the same set of addresses